File: matlab/ex_laplace.m
addpath('rchol/')

% 3D constant Poisson equation with dirichlet boundary
n = 128;
A = laplace_3d(n); % see 'rchol/laplace_3d.m'
fprintf('N: %d\n', size(A,1))

% random RHS
b = rand(size(A, 1), 1);

% compute preconditioner after reordering
tic
p = amd(A);
%p = 1:size(A,1);
G = rchol(A(p, p));
t1 = toc;
fprintf('fill-in ratio: %.2e\n', 2*nnz(G)/nnz(A))

% solve with PCG
tol = 1e-10;
maxit = 200;

tic
[x, flag, relres, itr] = pcg(A(p,p), b(p), tol, maxit, G, G');
t2 = toc;

fprintf('# CG iterations: %d\n', itr)

% verify solution
y = zeros(length(x), 1);
y(p) = x;
fprintf('Relative residual: %.2e\n', norm(b-A*y)/norm(b))
fprintf('Setup time: %.2f\n', t1)
fprintf('PCG time: %.2f\n', t2)

File: python/ex_laplace.py
import numpy as np 
from numpy.linalg import norm
from rchol import *
from util import *


# SDDM matrix from 3D constant Poisson equation
n = 20
A = laplace_3d(n)

# random RHS
N = A.shape[0]
b = np.random.rand(N)

# compute preconditioner after reordering (single thread)
'''
p = np.random.permutation(N)
Aperm = A[p[:,None],p]
G = rchol(Aperm)
print('fill-in ratio: {:.2}'.format(2*G.nnz/A.nnz))
'''

# compute preconditioner after reordering (multi thread)
p, G = rchol(A, 2)
Aperm = A[p[:, None], p]
print('fill-in ratio: {:.2}'.format(2*G.nnz/A.nnz))


# solve with PCG
tol = 1e-6
maxit = 200
x, relres, itr = pcg(Aperm, b[p], tol, maxit, G, G.transpose().tocsr())
print('# CG iterations: {}'.format(itr))

# verify solution
y = np.zeros(N)
y[p] = x
print('Relative residual: {:.2e}'.format(norm(A*y-b)/norm(b)))


